mono/eglib/gstr.c

```diff
@@ -841,26 +849,50 @@ g_ascii_strup (const gchar *str, gssize len)
 	
 	ret = g_malloc (len + 1);
 	for (i = 0; i < len; i++)
-		ret [i] = (guchar) g_ascii_toupper (str [i]);
+		ret [i] = g_ascii_toupper (str [i]);
 	ret [i] = 0;
 	
 	return ret;
 }
 
+static
+int
+g_ascii_charcmp (char c1, char c2)
+{
+	// Do not subtract, to avoid overflow.
+	// Use unsigned to mimic strcmp, and so
+	// shorter strings compare as less.
+
+	const guchar u1 = (guchar)c1;
+	const guchar u2 = (guchar)c2;
+	return (u1 < u2) ? -1 : (u1 > u2) ? 1 : 0;
+}
+
+static
+int
+g_ascii_charcasecmp (char c1, char c2)
+{
+	return g_ascii_charcmp (g_ascii_tolower (c1), g_ascii_tolower (c2));
+}
+
 gint
 g_ascii_strncasecmp (const gchar *s1, const gchar *s2, gsize n)
 {
+	// Unlike strncmp etc. this function does not stop at nul,
+	// unless there is a mismatch.
+
+	if (s1 == s2)
+		return 0;
+
 	gsize i;
-	
+
 	g_return_val_if_fail (s1 != NULL, 0);
 	g_return_val_if_fail (s2 != NULL, 0);
 
 	for (i = 0; i < n; i++) {
-		gchar c1 = g_ascii_tolower (*s1++);
-		gchar c2 = g_ascii_tolower (*s2++);
-		
-		if (c1 != c2)
-			return c1 - c2;
```

Comment:
Did anything rely on the subtraction? Should check the callers. This is a significant change to function behavior, though I think it's a reasonable one.

Reply:
Highly unlikely.
The usual idiom is <0, 0, or >0 and the actual values <0 and >0 are not important.

Reply:
Subtract is often-not-always incorrect due to overflow and hopefully nobody depended on incorrect results there.

Reply:
There are tests that assume the behavior and should be updated, though I don't see anything else (and am unsure of whether or not these tests are ever actually run). https://github.com/mono/mono/blob/ab69814598fbdc5fb66c59a0709eda2a61c3511f/mono/eglib/test/string-util.c#L627

Reply:
Wow that is a bad test. I've never seen anyone depend on other than <0, 0, >0 for "compare" for `strcmp`, `strncasecmp`, `_stricmp`, and that is all they guarantee. The actual values are not guaranteed. (See
https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/stricmp-wcsicmp-mbsicmp-stricmp-l-wcsicmp-l-mbsicmp-l?view=vs-2019, http://pubs.opengroup.org/onlinepubs/009695399/functions/strncmp.html, http://pubs.opengroup.org/onlinepubs/009695399/functions/strcmp.html http://pubs.opengroup.org/onlinepubs/009695399/functions/strcasecmp.html
etc.).

This is a custom mono routine so it could have a more "precise" but bad interface, but it would no longer resemble what it is based on and people could not switch back and forth.

There could be some sort of tie-breaking when chars vary only in case, but I don't think the old code worked that way. And things like qsort/bsearch would not notice.

I have seen the overflow bug repeatedly, for example comparing a GUID a ULONG at a time. The bug only shows up when all the bits are actually used. For example if the text was strictly ASCII and not UTF8, it'd be ok. But this seems like too much trouble to understand and code, vs. doing simple compares. And while it says "ascii", we'd have to look more closely to rule out UTF8 or other. I think "ASCII" might be more the algorithm and not the data.

Reply:
I believe the glib tests are not run. I made some improvements there, i.e. fixed them to be easily built, but stopped there.

Reply:
Yeah, the test definitely isn't great. I guess if they're not being run it doesn't matter too much, though in lieu of a fixed test a note would be nice.

Reply:
The test should be fixed with this PR. I hadn't noticed it.
People should occasionaly run these tests. I didn't. And confirm it fails.
We should probably enable the tests for real too.